docs/architecture/adr-005-subscription.md

```diff
@@ -0,0 +1,225 @@
+
+## Status
+
+#### Proposed
+
+## Implementation Changes
+
+Introduce the following messages to `x/subscription` module:
+
+```go
+// CreateSubscriptionMsg allows Service Provider to create new subscription service.
+type CreateSubscriptionMsg struct {
+    Name      string         // unique, human-readable name for subscription service
+    Amount    sdk.Coins      // amount to be collected for each subscription period
+    Period    time.time      // duration of subscription period
+    Collector sdk.AccAddress // address that will collect subscription payments
+}
+```
+
+```go
+// SubscribeMsg allows subscriber account to subscribe to provided service
+// If subscription to the service from user already exists, this msg is treated as renewal.
+// If msg.Limit != -1, then subscription.Limit += msg.Limit. Else, subscription.Limit = -1 (unlimited)
+type SubscribeMsg struct {
+    Name       string         // name of service to subscribe to
+    Subscriber sdk.AccAddress // address of subscriber
+    Limit      int64          // Maximum number of periods that subscription remains active. Limit = -1 implies no limit
+}
+```
+
+```go
+// UnsubscribeMsg allows subscriber to inactivate an active subscription
+type UnsubscribeMsg struct {
+    Name       string         // name of service to unsubscribe to
+    Subscriber sdk.AccAddress // address of subscriber
+}
+```
+
+```go
+// CollectMsg allows a Collector for a service to collect payments on due subscriptions that are processed off a FIFO queue.
+type CollectMsg struct {
+    Name      string         // name of subscription service to collect payments from
+    Collector sdk.AccAddress // address that will collect payments
+    Limit     int64          // maximum number of items to process in FIFO duequeue. If Limit = -1, try to process all due subscriptions
+}
+```
+
+Create a new store with the following key-values:
+
+`Address => []SubscriptionID // List of active/inactive subscriptions owned by the users`
+
+`Metadata:{Name} => SubscriptionMetaData`
+
+`DueQueue:{Name} => []SubscriptionID // FIFO queue of subscriptions. CONTRACT: All due subscriptions exist before all undue subscriptions. All subscriptions in DueQueue are active`
+
+`SubscriptionID => Subscription`
+
+```go
+// SubscriptionID is a unique identifier to a subscription struct
+// hash(Name+Address)
+type SubscriptionID []byte
+```
+
+```go
+// Subscription Metadata contains information necessary for processing subscriptions
+type Metadata struct {
+    Amount    sdk.Coins      // amount to be collected for each subscription period
+    Period    time.time      // duration of subscription period
+    Collector sdk.AccAddress // address that will collect subscription payments
+}
+```
+
+```go
+// Represents a user's subscription to a service
+type Subscription {
+    Name       string
+    Subscriber sdk.AccAddress
+    Limit      int64
+    LastPaid   time.Time
+    Active     bool
```

Comment:
see the comment above. I think we can delete this field if we delete the subscription from the store once the limit was reached

Reply:
I'd like users to be able to query for subscriptions that have went inactive in the store and then choose to renew whatever inactive subscriptions they wish

Reply:
they can still query the subscription `Terms`

Reply:
+ you can still query the events generated by the transactions